test/DGmethods/compressible_Navier_Stokes/mms_bc.jl

```diff
@@ -23,116 +23,115 @@ else
   const ArrayTypes = (Array, )
 end
 
-const _nstate = 5
-const _ρ, _U, _V, _W, _E = 1:_nstate
-const stateid = (ρid = _ρ, Uid = _U, Vid = _V, Wid = _W, Eid = _E)
-const statenames = ("ρ", "U", "V", "W", "E")
 
-const _nviscstates = 6
-const _τ11, _τ22, _τ33, _τ12, _τ13, _τ23 = 1:_nviscstates
-
-const _ngradstates = 3
-const _states_for_gradient_transform = (_ρ, _U, _V, _W)
+struct AtmosModel <: BalanceLaw
+end
 
-if !@isdefined integration_testing
-  const integration_testing =
-    parse(Bool, lowercase(get(ENV,"JULIA_CLIMA_INTEGRATION_TESTING","false")))
+dimension(::AtmosModel) = 3
+vars_aux(::AtmosModel) = ()
+vars_state(::AtmosModel) = (:ρ, :ρu, :ρv, :ρw, :ρe)
+vars_state_for_transform(::AtmosModel) = (:ρ, :ρu, :ρv, :ρw)
+vars_transform(::AtmosModel) = (:u, :v, :w)
+vars_diffusive(::AtmosModel) = (:τ11, :τ22, :τ33, :τ12, :τ13, :τ23)
+
+function flux!(::AtmosModel, flux::Grad, state::State, diffstate::State, auxstate::State, t::Real)
+  # preflux
+  γ = γ_exact  
+  ρinv = 1 / state.ρ
+  u, v, w = ρinv * state.ρu, ρinv * state.ρv, ρinv * state.ρw
+  P = (γ-1)*(state.ρe - ρinv * (state.ρu^2 + state.ρv^2 + state.ρw^2) / 2)
+
+  # invisc terms
+  flux.ρ  = (state.ρu          , state.ρv          , state.ρw)
+  flux.ρu = (u * state.ρu  + P , v * state.ρu      , w * state.ρu)
+  flux.ρv = (u * state.ρv      , v * state.ρv + P  , w * state.ρv)
+  flux.ρw = (u * state.ρw      , v * state.ρw      , w * state.ρw + P)
+  flux.ρe = (u * (state.ρe + P), v * (state.ρe + P), w * (state.ρe + P))
+
+  # viscous terms
+  flux.ρu .-= (diffstate.τ11, diffstate.τ12, diffstate.τ13)
+  flux.ρv .-= (diffstate.τ12, diffstate.τ22, diffstate.τ23)
+  flux.ρw .-= (diffstate.τ13, diffstate.τ23, diffstate.τ33)
+
+  flux.ρe .-= (u * diffstate.τ11 + v * diffstate.τ12 + w * diffstate.τ13,
+               u * diffstate.τ12 + v * diffstate.τ22 + w * diffstate.τ23,
+               u * diffstate.τ13 + v * diffstate.τ23 + w * diffstate.τ33)
 end
 
-include("mms_solution_generated.jl")
+function transform!(::AtmosModel, transformstate::State, state::State, auxstate::State, t::Real)
+  ρinv = 1 / state.ρ
+  transformstate.u = ρinv * state.ρu,
+  transformstate.v = ρinv * state.ρv
+  transformstate.w = ρinv * state.ρw
+end
 
-# preflux computation
-@inline function preflux(Q, _...)
-  γ::eltype(Q) = γ_exact
-  @inbounds ρ, U, V, W, E = Q[_ρ], Q[_U], Q[_V], Q[_W], Q[_E]
-  ρinv = 1 / ρ
-  u, v, w = ρinv * U, ρinv * V, ρinv * W
-  ((γ-1)*(E - ρinv * (U^2 + V^2 + W^2) / 2), u, v, w, ρinv)
+function diffusive!(::AtmosModel, diff::State, ∇transform::Grad)
+  dudx, dudy, dudz = ∇transform.u
+  dvdx, dvdy, dvdz = ∇transform.v
+  dwdx, dwdy, dwdz = ∇transform.w
+
+  # strains
+  ϵ11 = dudx
+  ϵ22 = dvdy
+  ϵ33 = dwdz
+  ϵ12 = (dudy + dvdx) / 2
+  ϵ13 = (dudz + dwdx) / 2
+  ϵ23 = (dvdz + dwdy) / 2
+
+  # deviatoric stresses
+  diff.τ11 = 2μ * (ϵ11 - (ϵ11 + ϵ22 + ϵ33) / 3)
```

Comment:
We wanted to define `\tau` as momentum flux tensor, to be consistent with other diffusive (scalar) fluxes (see CLIMA-atmos.tex). So then `\tau = - 2\nu (\epsilon - tr(eps)/3)I_3` (note minus sign), where `\nu` is the kinematic viscosity (rather than the dynamic viscosity `\mu` above).